### PR TITLE
[DataPipe] Ensures Prefetcher shuts down properly

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #1166

Differential Revision: [D45967152](https://our.internmc.facebook.com/intern/diff/D45967152)

### CODE REVIEW

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45967152).

File: torchdata/datapipes/iter/util/prefetcher.py [hunk 1]
```diff
@@ -152,6 +150,17 @@ def resume(self):
             self.prefetch_data.run_prefetcher = True
             self.prefetch_data.paused = False
 
+    @final
+    def shutdown(self):
+        if self.prefetch_data:
+            self.prefetch_data.run_prefetcher = False
+            self.prefetch_data.stop_iteration = True
+        if self.thread:
+            self.thread.join()
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/prefetcher.py:159]: We need to make sure this function is called when new `iterator` is created. So, you might want to add the following lines to `__iter__`:
```py
if self.thread is not None:
    self.shutdown()
```

**Review by ejguan — APPROVED** [review]
Overall LGTM with a comment below

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45967152).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45967152).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45967152).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45967152).

File: torchdata/datapipes/iter/util/prefetcher.py (hunk 1 above)
Comment (NivekT) [reply torchdata/datapipes/iter/util/prefetcher.py]: I actually think it is handled by `reset` method, which is called whenever `__iter__` is called

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45967152).

**Review by ejguan — APPROVED** [review]
Super nit: Can we do `self.thread = Thread()` in `__iter__`?

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45967152).

**Pull request closed without merging**